### ISSUE TITLE
Refactoring 

### ISSUE THREAD

**Issue by j4ustin** [plaid/plaid-ruby#35]
The gem is about 4 months old now without a good refactoring. Add small, non-breaking issues below so they can be attended to.

**Comment by j4ustin** [comment]
@hosh Refer all refactoring issues to this thread. Thanks!

**Comment by hosh** [comment]
https://github.com/plaid/plaid-ruby/commit/d8fbb47727562150ed03c8eb0c814c534c5a2dac#commitcomment-10594966

**Comment by hosh** [comment]
https://github.com/plaid/plaid-ruby/commit/d8fbb47727562150ed03c8eb0c814c534c5a2dac#

**Comment by hosh** [comment]
https://github.com/plaid/plaid-ruby/commit/d8fbb47727562150ed03c8eb0c814c534c5a2dac#commitcomment-10595001

**Comment by hosh** [comment]
I'm tempted to say combine `new` and `update_account` methods since they are the exact same code. https://github.com/plaid/plaid-ruby/blob/master/lib/plaid/user/account/account.rb#L10-L32 But I think there are deeper issues that are not addressed.

  1. Why `#new` at all? I noticed that there are patterns like https://github.com/plaid/plaid-ruby/blob/master/lib/plaid.rb#L24 for all the objects. My guess is that the intent is to initialize a new `User`, `Account`, or `Transaction` object based on results returned from the http request. If that is so, then the Ruby way of doing this is not to define `#new` but to define `#initialize`: http://www-rohan.sdsu.edu/doc/ruby/chp_04/classes.html

     So in the above example, we would have:
     ```
     # Builds the user object and returns on successful authentication
    def user(res,api_level=nil)
      @user = Plaid::User.new(res, api_level)
    end
     ```

     by defining:
     ```
     class Plaid::User
       def initialize(res, api_level=[])
          # what was inside new
       end
     ```

**Comment by hosh** [comment]
Oh missed this https://github.com/plaid/plaid-ruby/blob/master/lib/plaid.rb#L31

So based on my reading of the intent of the code, don't need to `<<` to push `api_level` into `@user.permissions`. Rather you should be able to just do `@user.permissions = api_level`. 

However, you would use `<<` to push `api_level` if you are checking if those auth levels are available. But that might be beyond the scope of this gem.

**Comment by j4ustin** [comment]
@hosh Good catch! Just added that one today. Rushed, it works but it's unnecessary.

**Comment by hosh** [comment]
I see this pattern in several places: https://github.com/plaid/plaid-ruby/blob/master/lib/plaid/institution/institution.rb#L25

While it is valid Ruby, normally, we'd do:
```
    def build_institution(inst)
      self.id = inst['id']
      self.name = inst['name']
      self.type = inst['type']
      self.has_mfa = inst['has_mfa']
      self.mfa = inst['mfa']
    end
```

The `,` is usually used like this:
```
def foo
  [1, 2]
end
a, b = foo

puts a
puts b
# => 1
# => 2
```
So to my eyes, seeing assignments like that looks weird in Ruby since I'm expecting multiple assignments but that is not what is happening there.

**Comment by hosh** [comment]
@j4ustin cool, glad you're finding this helpful!

**Comment by j4ustin** [comment]
@hosh 

Would switching ```#new``` to ```#initialize``` break the existing functionality? 

Also, I'm finding this super helpful! I switch between a bunch of different languages in my day job, so (as I'm sure you're seeing) there is lots of syntax that may not be the Ruby way to do things. It's nice to have another set of eyes on this. Thanks for the help!

**Comment by hosh** [comment]
@j4ustin Cool, yeah I figured you switch around on languages a lot.

On the `new` and `initialize`. So usually, we mark out "internal" and "public" methods. The public methods are the ones you decide to support for users, while being able to change any of the internals in any way you wish. If a user decides to rely on internal methods, it's on them if things break. So this probably needs to be clarified.

Let's say we are only supporting what is in the `README`. Meaning, if you want a `Plaid::User`, you are going to get it by calling `Plaid.add_user`, `Plaid.set_user` and `Plaid.existing_user`. (And maybe not even `existing_user` since that is being called by `set_user`). You would want to have the getter methods be public.

This makes sense since, as an dev writing an app using Plaid, I really don't care about building my own instance of `Plaid::User`. I want to grab my end-user's bank auth info, send it to this code, and get back a bunch of data accessible by getter methods.

It would be really weird if I already have code that would explicitly call `Plaid::User.new` (at least, based on how you designed the gem).

Maybe some other community members can comment? Or maybe play it safe and just work all of this towards a `2.0` gem, explicitly define what are public methods (that the gem is guaranteeing backwards compatibility for minor releases), and what are internal methods with no such guarantee.

There's actually other stuff I'd like to send more PRs for, but I'm glad we're having this discussion!

**Comment by j4ustin** [comment]
@hosh Yep thats right. 

I'd like to hear what other people think. It seems like there should only be a handful of methods that we should support publicly, and almost all of them are on the Plaid object or an initialized user object. I'll add these to the read me. 

The approach I'd like to take is 1) fix all non-breaking issues in the V1 gem, then 2) shift the rest of the issues into V2 and prep that for a release. We can small issues and mark them using milestones. Then solve with PR's.

**Comment by hosh** [comment]
All right, fair enough.

**Comment by hosh** [comment]
Many of these are addressed by #42
